pyop2/compilation.py

```diff
@@ -409,40 +422,50 @@ def __init__(self, cppargs=[], ldargs=[], cpp=False, comm=None):
 
 
 @collective
-def load(src, extension, fn_name, cppargs=[], ldargs=[],
+def load(jitmodule, extension, fn_name, cppargs=[], ldargs=[],
          argtypes=None, restype=None, compiler=None, comm=None):
     """Build a shared library and return a function pointer from it.
 
-    :arg src: A string containing the source to build
+    :arg jitmodule: The JIT Module which can generate the code to compile, or
+        the string representing the source code.
     :arg extension: extension of the source file (c, cpp)
     :arg fn_name: The name of the function to return from the resulting library
     :arg cppargs: A list of arguments to the C compiler (optional)
     :arg ldargs: A list of arguments to the linker (optional)
-    :arg argtypes: A list of ctypes argument types matching the
-         arguments of the returned function (optional, pass ``None``
-         for ``void``).
+    :arg argtypes: A list of ctypes argument types matching the arguments of
+         the returned function (optional, pass ``None`` for ``void``). This is
+         only used when string is passed in instead of JITModule.
```

Comment:
So the reason for this code churn is that the code can now either be something that makes code (via wrapper generation), or the string itself. So perhaps we should pass in an object that can return a cache key and the code in both cases. Then this change would be smaller.


e.g.
```
class StrCode(object):
    def __init__(self, code):
       self.code_to_compile = code
    @cached_property
    def cache_key(self):
       return md5(self.code_to_compile.encode())
```